Complete CLI docs. (#1029)
quickwit-inc/quickwit: docs/reference/cli.md

sidebar_position: 1
---

> Quickwit command line tool lets you create, ingest, search, start search and indexer servers. For configuration, `quickwit` needs a [config file path](config.md) that you can configure with `QW_CONFIG` environment variable: `export QW_CONFIG=./config/quickwit.yaml`.

Reviewer: ```suggestion
Quickwit command line tool lets you create, ingest, search, start search and indexer servers. For configuration, `quickwit` needs a [config file path](config.md) that you can specify with `QW_CONFIG` environment variable: `export QW_CONFIG=./config/quickwit.yaml`.
```

`quickwit <command name> --help` displays the documentation for the command and a usage example.

> ## Environment Variables

Reviewer 2: In the last review you asked to put the QW_CONFIG env at the top, since it simpliyfies the whole usage. I think that makes sense

Author: Yes indeed, but I only wanted to talk about `QW_CONFIG` which is the most important and leave other env variables at the end.

### Version

`quickwit --version` displays the version. It is helpful for reporting bugs.

### Syntax

The CLI is structured into high-level commands with subcommands.
`quickwit [command] [subcommand] [args]`.

* `command`: `index`, `split`, `source` and `service`. 

> [comment]: <> (Insert auto generated CLI docs from here.)

Reviewer 2: is this ignored by docusaurus?

Author: Hope so, will see that very soon.

*Examples*

> *Create a new index with the default local metastore.*

Author: This is not exact.

```bash

> quickwit index create --index wikipedia --index-config wikipedia_index_config.yaml  --config=./config/quickwit.yaml

Reviewer 2: I prefer it without subfolder since it's simpler for the user
`--config=quickwit.yaml`

Author: I assume the user have a fresh install of quickwit and thus has the YAML file in the config dir. I think both are ok.

Reviewer 2: I don't understand how a new install is related to having the config file in the subfolder

quickwit index create --index-config-uri ./wikipedia_index_config.yaml  --config=quickwit.yaml
quickwit index create --index wikipedia --index-config wikipedia_index_config.yaml  --config=./config/quickwit.yaml
```

### index ingest

Indexes a dataset consisting of newline-delimited JSON objects located at `input-path` or read from *stdin*. 
The data is appended to the target index specified by `index` unless `overwrite` is passed. `input-path` can be a file or another command output piped into stdin. 
Currently, only local datasets are supported. 
By default, quickwit's indexer will work with a heap of 2 GiB of memory, but this can be set with the `heap-size` option in the indexing_settings.resources section in the config. 

> This does not directly reflect the overall memory usage of `quickwit index ingest`, but doubling this value should give a fair approximation.

Reviewer 2: this is useful information, or is it outdated?

Author: I prefer to have the details on the index config page to avoid duplication. Though the writing is not really good here, I will improve it.

`--index` ID of the target index.    
`--config` Quickwit config file.    

*Examples*

*Displays descriptive statistics of your index*
```bash

> quickwit index describe --index wikipedia --config ./config/quickwit.yaml

Reviewer 2: Not sure, but we could add example output?

```suggestion
quickwit index describe --index wikipedia --config ./config/quickwit.yaml

1. General infos
===============================================================================
Index id:                           wikipedia
Index uri:                          file:///home/quickwit-indices/qwdata/indexes/wikipedia
Number of published splits:         1
Number of published documents:      300000
Size of published splits:           448 MB

2. Statistics on splits
===============================================================================
Document count stats:
Mean ± σ in [min … max]:            300000 ± 0 in [300000 … 300000]
Quantiles [1%, 25%, 50%, 75%, 99%]: [300000, 300000, 300000, 300000, 300000]

Size in MB stats:
Mean ± σ in [min … max]:            448 ± 0 in [448 … 448]
Quantiles [1%, 25%, 50%, 75%, 99%]: [448, 448, 448, 448, 448]

```

Author: good idea

*Options*

`--index` ID of the target index.    
`--config` Quickwit config file.    

> `--query` Query expressed in natural query language (barack AND obama) OR "president of united states").    

Reviewer 2: I think we need a link here or reference tantivy syntax, or we let the user guess the actual syntax

Author: I put a link to the query language in the description.

Reviewer 2: I think it would be good to have it also in the cli for users that explore the cli without the docs

Author: Will do.

`quickwit service run [args]`
### service run searcher

Starts a web server at `rest_listing_address:rest_list_port` that exposes the [Quickwit REST API](search-api.md)

> where `rest_listing_address` and `rest_list_port` are defined in Quickwit config file (quickwit.yaml).

Reviewer 2: ```suggestion
where `listen_address` and `rest_listen_port` are defined in the Quickwit config file (quickwit.yaml).
```

*Examples*

*Start a Searcher*
```bash
quickwit service run searcher --config=./config/quickwit.yaml
```

*curl request and specify fields to search*
```bash
curl -L "http://127.0.0.1:7280/api/v1/wikipedia/search/stream?query=clinton&searchField=body,title" 
```

*curl request stream fast field, HTTP1.1 chunked transfer encoding*
```bash
curl -L "http://127.0.0.1:7280/api/v1/wikipedia/search/stream?query=tangkhul&searchField=body,title&fastField=number&outputFormat=csv"  
```

*curl request stream fast field as binary, force HTTP2*
```bash
curl -L --http2-prior-knowledge "http://127.0.0.1:7280/api/v1/myindex/search/stream?query=clinton&searchField=body,title&fastField=number&outputFormat=clickHouseRowBinary"  
```

### service run indexer

Starts an indexing server that consumes the sources of index IDs passed in `--indexes` argument.

> If all sources of all indexes are exhausted, the server will stop.

Reviewer 2: Last time I checked the server did just idle forever

Author: We need to check that :/

Reviewer 2: Yes it's still the case

Author: Yes, I read the code and the expected behavior is to wait until the `IndexingServer` actor exits. And this will not happen unless there is an error happens when processing a message. I'm removing the sentence.

*Examples*

*Start an Indexer*
```bash

> quickwit service run indexer --config=./config/quickwit.yaml

Reviewer 2: I prefer the examples to be more complete, and indexer without sources doesn't do anything

Author: good point, I will keep it.

`--type` Type of the source. Available types are: `file` and `kafka`.    
`--params` Parameters for the source formatted as a JSON object passed inline or via a file. Parameters are source-specific. Please, refer to the source's documentation for more details.    
`--config` Quickwit config file.    

*Examples*

*Add a file source to `wikipedia` index*
```bash
 
quickwit source add --index wikipedia --id wikipedia-source --type file --params '{"filepath":"wiki-articles.json"}'

```

*Add a Kafka source to `wikipedia` index*
```bash
 
cat << EOF > wikipedia-source.json
{
  "topic": "wikipedia",
  "client_params": {
    "bootstrap.servers": "localhost:9092",
    "group.id": "my-group-id",
    "security.protocol": "SSL"
  }
}
EOF

> quickwit source add --index wikipedia --id wikipedia-source --type kafka --params wikipedia-source.json

Reviewer 2: I think replacing `id` with `source` would be more consistent
e.g.
`quickwit index create --index wikipedia` (not id)

Author: I had this discussion with @guilload but we decided to not touch it for now
